### PR TITLE
Combine: Move SqlEnvChange to shared and sync usage

### PR DESCRIPTION
**Pull request by Wraith2**
As the title says. Makes SqlEnvChange shared and changes netfx to use the same logic as netcore which has better a better memory allocation profile. Later on after this and https://github.com/dotnet/SqlClient/pull/1199 are merged the SqlEnvChange objects can be pooled which will further reduce allocations on reconnection and query.

### CODE REVIEW

**Review by cheenamalhotra — APPROVED** [review]

**Review by johnnypham — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlEnvChange.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        internal byte[] oldBinValue;
+        internal long newLongValue;
+        internal long oldLongValue;
+        internal SqlCollation newCollation;
+        internal SqlCollation oldCollation;
+        internal RoutingInfo newRoutingInfo;
+        internal bool newBinRented;
+        internal bool oldBinRented;
+
+        internal SqlEnvChange Next;
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlEnvChange.cs:34]: Could you fix the naming violations, plz?

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 2]
```diff
@@ -2941,39 +2947,38 @@ internal bool TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataRead
...
-        private bool TryProcessEnvChange(int tokenLength, TdsParserStateObject stateObj, out SqlEnvChange[] sqlEnvChange)
+        private bool TryProcessEnvChange(int tokenLength, TdsParserStateObject stateObj, out SqlEnvChange sqlEnvChange)
         {
             // There could be multiple environment change messages following this token.
             byte byteLength;
             int processedLength = 0;
-            int nvalues = 0;
-            SqlEnvChange[] envarray = new SqlEnvChange[3];  // Why is this hardcoded to 3?
+            //int nvalues = 0;
+            //SqlEnvChange[] envarray = new SqlEnvChange[3];  // Why is this hardcoded to 3?
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs:2956]: Could be removed if you didn't leave them by purpose.
```suggestion
```

**Review by DavoudEshtehari — DISMISSED** [review]

**Comment by Wraith2** [comment]
merge conflicts and feedback fixed.

**Comment by Wraith2** [comment]
I went through the original commits from corefx and I think I've found the problem and corrected it. I can't debug the DTC related tests locally though because I can't get DTC working (spent my friday night trying) so if this doesn't work either I'll need some help or we'll have to close and retry this another time.

**Comment by johnnypham** [comment]
> I went through the original commits from corefx and I think I've found the problem and corrected it. I can't debug the DTC related tests locally though because I can't get DTC working (spent my friday night trying) so if this doesn't work either I'll need some help or we'll have to close and retry this another time.

What have you tried? Adding a firewall rule for msdtc.exe worked for me.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj [hunk 3]
```diff
@@ -569,6 +569,9 @@
     <Compile Include="..\..\src\Microsoft\Data\SqlClient\SqlEnclaveSession.cs">
       <Link>Microsoft\Data\SqlClient\SqlEnclaveSession.cs</Link>
     </Compile>
+    <Compile Include="..\..\src\Microsoft\Data\SqlClient\SqlEnvChange.cs">
+      <Link>Microsoft\Data\SqlClient\SqlEnvChange.cs</Link>
+    </Compile>
```
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft.Data.SqlClient.csproj]: Done.

**Comment by Wraith2** [comment]
This is another PR which I'm going to have to re-do because it's been open so long the rebase can't be done cleanly. There are errors around Server namespace types.

**Pull request closed without merging**